firmware/src/sensor_model.c

```diff
@@ -0,0 +1,234 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ * sensor_model.c
+ * Copyright (C) 2015 Simon Newton
+ */
+#include "sensor_model.h"
+
+#include <stdlib.h>
+
+#include "coarse_timer.h"
+#include "constants.h"
+#include "rdm_frame.h"
+#include "rdm_responder.h"
+#include "rdm_util.h"
+#include "utils.h"
+
+// Various constants
+#define SOFTWARE_VERSION 0x00000000
+#define SENSOR_SAMPLE_RATE 10000
+
+enum { NUMBER_OF_SENSORS = 3 };
+
+static const char DEVICE_MODEL_DESCRIPTION[] = "Ja Rule Sensor Device";
+static const char SOFTWARE_LABEL[] = "Alpha";
+static const char DEFAULT_DEVICE_LABEL[] = "Ja Rule";
+static const char SENSOR_NAME1[] = "Temperature";
+static const char SENSOR_NAME2[] = "Missing Sensor";
+static const char SENSOR_NAME3[] = "Voltage";
+
+static const ResponderDefinition RESPONDER_DEFINITION;
+
+/*
+ * @brief The sensor model state.
+ */
+typedef struct {
+  CoarseTimer_Value sensor_sample_time;
+  SensorData sensors[NUMBER_OF_SENSORS];
+} SensorModel;
+
+static SensorModel g_sensor_model;
+
+void SampleSensors() {
+  g_sensor_model.sensor_sample_time = CoarseTimer_GetTime();
+
+  unsigned int i = 0;
+  for (; i < NUMBER_OF_SENSORS; i++) {
+    int16_t new_value = (Random_PseudoGet() % (
+        RESPONDER_DEFINITION.sensors[i].range_maximum_value -
+        RESPONDER_DEFINITION.sensors[i].range_minimum_value)) +
+        RESPONDER_DEFINITION.sensors[i].range_minimum_value;
+    g_sensor_model.sensors[i].present_value = new_value;
+    if (new_value < g_sensor_model.sensors[i].lowest_value) {
+      g_sensor_model.sensors[i].lowest_value = new_value;
+    }
+    if (new_value > g_sensor_model.sensors[i].highest_value) {
+      g_sensor_model.sensors[i].highest_value = new_value;
+    }
+  }
+}
+
+// Public Functions
+// ----------------------------------------------------------------------------
+void SensorModel_Initialize(const SensorModelSettings *settings) {}
+
+static void SensorModel_Activate() {
+  g_responder.def = &RESPONDER_DEFINITION;
+  unsigned int i = 0;
+  for (; i < NUMBER_OF_SENSORS; i++) {
+    g_sensor_model.sensors[i].present_value = 0;
+    g_sensor_model.sensors[i].lowest_value = 0;
+    g_sensor_model.sensors[i].highest_value = 0;
+    g_sensor_model.sensors[i].recorded_value = 0;
+    if (i == 1) {
+      // Sensor 1 always nacks with a HARDWARE_FAULT
```

Comment:
Clarify the comment regarding 0 based numbering, perhaps give the sensor name instead.